File: lib/utils/context_retrieval.dart
import 'package:ai_life_recorder/models/context_pack.dart';
import 'package:ai_life_recorder/models/memory.dart';
import 'package:ai_life_recorder/models/photo.dart';
import 'package:ai_life_recorder/models/task.dart';
import 'package:ai_life_recorder/models/summary.dart';

List<String> extractKeywords(String text) {
  if (text.isEmpty) return <String>[];
  final stopChars = <String>{
    '的',
    '了',
    '是',
    '我',
    '你',
    '他',
    '在',
    '这',
    '个',
    '们',
    '吧',
    '呢'
  };
  final seen = <String>{};
  final result = <String>[];
  bool hasCJK = text.runes.any((r) {
    return (r >= 0x4E00 && r <= 0x9FFF) || (r >= 0x3400 && r <= 0x4DBF);
  });
  if (hasCJK) {
    final buffer = StringBuffer();
    for (var rune in text.runes) {
      if ((rune >= 0x4E00 && rune <= 0x9FFF) ||
          (rune >= 0x3400 && rune <= 0x4DBF)) {
        buffer.writeCharCode(rune);
      } else {
        buffer.write(' ');
      }
    }
    final segments =
        buffer.toString().split(RegExp(r'\s+')).where((s) => s.isNotEmpty);
    for (var seg in segments) {
      if (seg.length < 2) continue;
      for (var i = 0; i <= seg.length - 2; i++) {
        final token = seg.substring(i, i + 2);
        bool containsStop =
            token.runes.any((r) => stopChars.contains(String.fromCharCode(r)));
        if (containsStop) continue;
        if (!seen.contains(token)) {
          seen.add(token);
          result.add(token);
        }
      }
    }
    return result;
  } else {
    final parts = text.split(RegExp(r'\s+'));
    for (var p in parts) {
      var w = p.trim();
      if (w.isEmpty) continue;
      w = w.replaceAll(RegExp(r'^[^A-Za-z]+'), '');
      w = w.replaceAll(RegExp(r'[^A-Za-z]+$'), '');
      if (w.length >= 2) {
        final lower = w.toLowerCase();
        if (!seen.contains(lower)) {
          seen.add(lower);
          result.add(lower);
        }
      }
    }
    return result;
  }
}

double timeScore(DateTime a, DateTime b) {
  final au = a.toUtc();
  final bu = b.toUtc();
  final ad = DateTime.utc(au.year, au.month, au.day);
  final bd = DateTime.utc(bu.year, bu.month, bu.day);
  final diff = ad.difference(bd).inDays.abs();
  if (diff == 0) return 2.0;
  if (diff == 1) return 1.5;
  if (diff == 2) return 1.0;
  if (diff <= 7) return 0.5;
  return 0.0;
}

double _countKeywordHits(String text, List<String> keywords) {
  if (text.isEmpty || keywords.isEmpty) return 0.0;
  final lower = text.toLowerCase();
  double count = 0.0;
  for (var kw in keywords) {
    if (kw.isEmpty) continue;
    final kwLower = kw.toLowerCase();
    int start = 0;
    while (true) {
      final idx = lower.indexOf(kwLower, start);
      if (idx == -1) break;
      count += 1.0;
      start = idx + kwLower.length;
    }
  }
  return count;
}

ContextPack buildContextPack({
  required String eventDescription,
  String? emotion,
  String? actionTaken,
  String? result,
  required List<String> tags,
  required List<Memory> memories,
  required List<Photo> photos,
  required List<Task> tasks,
  required List<Summary> summaries,
}) {
  final now = DateTime.now().toUtc();
  final combinedTextParts = <String>[];
  combinedTextParts.add(eventDescription);
  if (emotion != null && emotion.isNotEmpty) combinedTextParts.add(emotion);
  if (actionTaken != null && actionTaken.isNotEmpty) {
    combinedTextParts.add(actionTaken);
  }
  if (result != null && result.isNotEmpty) combinedTextParts.add(result);
  final combinedText = combinedTextParts.join(' ');
  final keywords = extractKeywords(combinedText);
  double computeScoreForTagsAndText(
      List<String> recordTags, String text, DateTime date) {
    final tagIntersection = tags.where((t) => recordTags.contains(t)).length;
    final tagScore = 3.0 * tagIntersection;
    final kwHits = _countKeywordHits(text, keywords);
    final kwScore = 2.0 * kwHits;
    final tScore = timeScore(now, date);
    return tagScore + kwScore + tScore;
  }

  final memScores = <Map<String, dynamic>>[];
  for (var m in memories) {
    final text = '${m.content} ${m.aiSummary ?? ''}'.trim();
    final score = computeScoreForTagsAndText(m.tags, text, m.createdAt);
    if (score > 0) {
      memScores.add({'score': score, 'memory': m});
    }
  }
  memScores
      .sort((a, b) => (b['score'] as double).compareTo(a['score'] as double));
  final memItems = <ContextItem>[];
  for (var i = 0; i < memScores.length && i < 5; i++) {
    final m = memScores[i]['memory'] as Memory;
    final title = (m.title != null && m.title!.trim().isNotEmpty)
        ? m.title!
        : (m.content.length <= 20 ? m.content : m.content.substring(0, 20));
    memItems.add(
        ContextItem(id: m.id, title: title, date: m.createdAt, tags: m.tags));
  }
  final photoScores = <Map<String, dynamic>>[];
  for (var p in photos) {
    if (!p.summaryConfirmed) continue;
    final text = (p.aiSummary ?? '');
    final score = computeScoreForTagsAndText(p.tags, text, p.takenAt);
    if (score > 0) {
      photoScores.add({'score': score, 'photo': p});
    }
  }
  photoScores
      .sort((a, b) => (b['score'] as double).compareTo(a['score'] as double));
  final photoItems = <ContextItem>[];
  for (var i = 0; i < photoScores.length && i < 5; i++) {
    final p = photoScores[i]['photo'] as Photo;
    final title = (p.aiSummary != null && p.aiSummary!.trim().isNotEmpty)
        ? p.aiSummary!
        : p.localPath;
    photoItems.add(
        ContextItem(id: p.id, title: title, date: p.takenAt, tags: p.tags));
  }
  final taskScores = <Map<String, dynamic>>[];
  for (var t in tasks) {
    final text = '${t.title} ${t.description ?? ''}'.trim();
    final date = t.createdAt;
    // Task 无 tags 字段，只按关键词+时间评分
    final score = computeScoreForTagsAndText(const <String>[], text, date);
    if (score > 0) {
      taskScores.add({'score': score, 'task': t});
    }
  }
  taskScores
      .sort((a, b) => (b['score'] as double).compareTo(a['score'] as double));
  final taskItems = <ContextItem>[];
  for (var i = 0; i < taskScores.length && i < 5; i++) {
    final t = taskScores[i]['task'] as Task;
    final title = t.title;
    taskItems.add(
        ContextItem(id: t.id, title: title, date: t.createdAt, tags: const []));
  }
  final sumScores = <Map<String, dynamic>>[];
  for (var s in summaries) {
    final textParts = <String>[];
    if (s.content != null) textParts.add(s.content!);
    if (s.themes.isNotEmpty) textParts.add(s.themes.join(' '));
    if (s.highlights.isNotEmpty) textParts.add(s.highlights.join(' '));
    final text = textParts.join(' ');
    final date = s.periodStart;
    final score = computeScoreForTagsAndText(s.themes, text, date);
    if (score > 0) {
      sumScores.add({'score': score, 'summary': s});
    }
  }
  sumScores
      .sort((a, b) => (b['score'] as double).compareTo(a['score'] as double));
  final summaryItems = <ContextItem>[];
  for (var i = 0; i < sumScores.length && i < 5; i++) {
    final s = sumScores[i]['summary'] as Summary;
    final title = (s.content != null && s.content!.trim().isNotEmpty)
        ? (s.content!.length <= 20 ? s.content! : s.content!.substring(0, 20))
        : (s.themes.isNotEmpty ? s.themes.join('、') : '');
    summaryItems.add(ContextItem(
        id: s.id, title: title, date: s.periodStart, tags: s.themes));
  }
  return ContextPack(
    memories: memItems,
    photos: photoItems,
    tasks: taskItems,
    summaries: summaryItems,
  );
}

File: src/utils/contextRetrieval.ts
// src/utils/contextRetrieval.ts
import { ContextItem, ContextPack, Memory, Photo, Task, Summary } from '../types';

export function extractKeywords(text: string): string[] {
  if (!text) return [];
  const stopChars = new Set(['的', '了', '是', '我', '你', '他', '在', '这', '个', '们', '吧', '呢']);
  const seen = new Set<string>();
  const result: string[] = [];

  let hasCJK = false;
  for (let i = 0; i < text.length; i++) {
    const code = text.charCodeAt(i);
    if ((code >= 0x4e00 && code <= 0x9fff) || (code >= 0x3400 && code <= 0x4dbf)) {
      hasCJK = true;
      break;
    }
  }

  if (hasCJK) {
    let cleaned = '';
    for (let i = 0; i < text.length; i++) {
      const code = text.charCodeAt(i);
      if ((code >= 0x4e00 && code <= 0x9fff) || (code >= 0x3400 && code <= 0x4dbf)) {
        cleaned += text[i];
      } else {
        cleaned += ' ';
      }
    }
    const segments = cleaned.split(/\s+/).filter(Boolean);
    for (const seg of segments) {
      if (seg.length < 2) continue;
      for (let i = 0; i <= seg.length - 2; i++) {
        const token = seg.substring(i, i + 2);
        if (Array.from(token).some((c) => stopChars.has(c))) continue;
        if (!seen.has(token)) {
          seen.add(token);
          result.push(token);
        }
      }
    }
    return result;
  } else {
    const parts = text.split(/\s+/);
    for (const p of parts) {
      const w = p.trim().replace(/^[^A-Za-z]+/, '').replace(/[^A-Za-z]+$/, '');
      if (w.length >= 2) {
        const lower = w.toLowerCase();
        if (!seen.has(lower)) {
          seen.add(lower);
          result.push(lower);
        }
      }
    }
    return result;
  }
}

export function timeScore(aDate: string | Date, bDate: string | Date): number {
  const a = typeof aDate === 'string' ? new Date(aDate) : aDate;
  const b = typeof bDate === 'string' ? new Date(bDate) : bDate;
  const aDay = Math.floor(a.getTime() / (1000 * 60 * 60 * 24));
  const bDay = Math.floor(b.getTime() / (1000 * 60 * 60 * 24));
  const diff = Math.abs(aDay - bDay);
  if (diff === 0) return 2.0;
  if (diff === 1) return 1.5;
  if (diff === 2) return 1.0;
  if (diff <= 7) return 0.5;
  return 0.0;
}

function countKeywordHits(text: string, keywords: string[]): number {
  if (!text || keywords.length === 0) return 0;
  const lower = text.toLowerCase();
  let count = 0;
  for (const kw of keywords) {
    if (!kw) continue;
    const kwLower = kw.toLowerCase();
    let start = 0;
    while (true) {
      const idx = lower.indexOf(kwLower, start);
      if (idx === -1) break;
      count += 1.0;
      start = idx + kwLower.length;
    }
  }
  return count;
}

export function buildContextPack(params: {
  eventDescription: string;
  emotion?: string | null;
  actionTaken?: string | null;
  result?: string | null;
  tags: string[];
  memories: Memory[];
  photos: Photo[];
  tasks: Task[];
  summaries: Summary[];
}): ContextPack {
  const { eventDescription, emotion, actionTaken, result, tags, memories, photos, tasks, summaries } = params;
  const now = new Date();

  const combinedTextParts = [eventDescription];
  if (emotion) combinedTextParts.push(emotion);
  if (actionTaken) combinedTextParts.push(actionTaken);
  if (result) combinedTextParts.push(result);
  const combinedText = combinedTextParts.join(' ');
  const keywords = extractKeywords(combinedText);

  function computeScore(recordTags: string[], text: string, date: string | Date): number {
    const tagIntersection = tags.filter((t) => recordTags.includes(t)).length;
    const tagScore = 3.0 * tagIntersection;
    const kwHits = countKeywordHits(text, keywords);
    const kwScore = 2.0 * kwHits;
    const tScore = timeScore(now, date);
    return tagScore + kwScore + tScore;
  }

  // Memories
  const memScores: { score: number; memory: Memory }[] = [];
  for (const m of memories) {
    const text = `${m.content} ${m.aiSummary || ''}`.trim();
    const score = computeScore(m.tags, text, m.createdAt);
    if (score > 0) {
      memScores.push({ score, memory: m });
    }
  }
  memScores.sort((a, b) => b.score - a.score);
  const memItems: ContextItem[] = memScores.slice(0, 5).map(({ memory: m }) => ({
    id: m.id,
    type: 'memory',
    titleOrSummary: m.title?.trim() || (m.content.length <= 20 ? m.content : m.content.substring(0, 20)),
    date: m.createdAt,
    tags: m.tags,
  }));

  // Photos (only confirmed)
  const photoScores: { score: number; photo: Photo }[] = [];
  for (const p of photos) {
    if (!p.summaryConfirmed) continue;
    const text = p.aiSummary || '';
    const score = computeScore(p.tags, text, p.takenAt);
    if (score > 0) {
      photoScores.push({ score, photo: p });
    }
  }
  photoScores.sort((a, b) => b.score - a.score);
  const photoItems: ContextItem[] = photoScores.slice(0, 5).map(({ photo: p }) => ({
    id: p.id,
    type: 'photo',
    titleOrSummary: p.aiSummary?.trim() || p.locationName || '照片记录',
    date: p.takenAt,
    tags: p.tags,
  }));

  // Tasks
  const taskScores: { score: number; task: Task }[] = [];
  for (const t of tasks) {
    const text = `${t.title} ${t.description || ''}`.trim();
    const score = computeScore([], text, t.createdAt);
    if (score > 0) {
      taskScores.push({ score, task: t });
    }
  }
  taskScores.sort((a, b) => b.score - a.score);
  const taskItems: ContextItem[] = taskScores.slice(0, 5).map(({ task: t }) => ({
    id: t.id,
    type: 'task',
    titleOrSummary: t.title,
    date: t.createdAt,
    tags: [],
  }));

  // Summaries
  const sumScores: { score: number; summary: Summary }[] = [];
  for (const s of summaries) {
    const themeNames = s.themes?.map((th) => th.name) || [];
    const text = `${s.content} ${themeNames.join(' ')} ${s.highlights?.join(' ') || ''}`.trim();
    const score = computeScore(themeNames, text, s.periodStart);
    if (score > 0) {
      sumScores.push({ score, summary: s });
    }
  }
  sumScores.sort((a, b) => b.score - a.score);
  const summaryItems: ContextItem[] = sumScores.slice(0, 5).map(({ summary: s }) => {
    const themeNames = s.themes?.map((th) => th.name) || [];
    const title = s.content?.trim()
      ? (s.content.length <= 20 ? s.content : s.content.substring(0, 20))
      : (themeNames.join('、') || `${s.type}度总结`);
    return {
      id: s.id,
      type: 'summary',
      titleOrSummary: title,
      date: s.periodStart,
      tags: themeNames,
    };
  });

  return {
    memories: memItems,
    photos: photoItems,
    tasks: taskItems,
    summaries: summaryItems,
  };
}
